rust/datafusion/src/scheduler/mod.rs

```diff
@@ -0,0 +1,381 @@
+        ));
+    }
+
+    let mut shuffle_location_map: HashMap<ShuffleId, ExecutorMeta> = HashMap::new();
+
+    let mut stage_status_map = HashMap::new();
+
+    for stage in &job.stages {
+        let stage = stage.borrow_mut();
+        stage_status_map.insert(stage.id, StageStatus::Pending);
+    }
+
+    // loop until all stages are complete
+    let mut num_completed = 0;
+    while num_completed < job.stages.len() {
+        num_completed = 0;
+
+        //TODO do stages in parallel when possible
+        for stage in &job.stages {
+            let stage = stage.borrow_mut();
+            let status = stage_status_map.get(&stage.id).unwrap();
+            match status {
+                StageStatus::Pending => {
+                    // have prior stages already completed ?
+                    if stage.prior_stages.iter().all(|id| {
+                        match stage_status_map.get(id) {
+                            Some(StageStatus::Completed) => true,
+                            _ => false,
+                        }
+                    }) {
+                        println!("Running stage {}", stage.id);
+                        let plan = stage
+                            .plan
+                            .as_ref()
+                            .expect("all stages should have plans at execution time");
+
+                        let stage_start = Instant::now();
+
+                        let exec = plan;
+                        let parts = exec.output_partitioning().partition_count();
+
+                        // build queue of tasks per executor
+                        let mut next_executor_id = 0;
+                        let mut executor_tasks: HashMap<String, Vec<ExecutionTask>> =
+                            HashMap::new();
+                        #[allow(clippy::needless_range_loop)]
+                        for i in 0..executors.len() {
+                            //executor_tasks.insert(executors[i].id.clone(), vec![]);
+                        }
+                        for partition in 0..parts {
+                            let task = ExecutionTask::new(
+                                job.id,
+                                stage.id,
+                                partition,
+                                plan.clone(),
+                                shuffle_location_map.clone(),
+                            );
+
+                            // load balance across the executors
+                            let executor_meta = &executors[next_executor_id];
+                            next_executor_id += 1;
+                            if next_executor_id == executors.len() {
+                                next_executor_id = 0;
+                            }
+
+                            let queue = executor_tasks
+                                .get_mut(&executor_meta.id)
+                                .expect("executor queue should exist");
+
+                            queue.push(task);
```

Comment:
I wonder how this will handle data flowing in a plan where a producer (eg scan) can make data faster than the consumer (eg aggregate) can consume it.

Reply:
Within each query stage, it is still a pull model. The leaf query stages will typically be a scan wrapped by some other operators. With the currently supported DataFusion operators, the `MergeExec` is essentially the shuffle exchange that would represent query stage boundaries. For example, `LocalLimit` and `HashAggregate(mode=Final)` operators would not start executing until their input query stages complete. This is not dissimilar to how `MergeExec` works today in that it blocks on the input threads, but it is occupying a thread itself while it waits in the current model but with the scheduler model, the `MergeExec` (possibly replaced with `ShuffleReader`) would not start running until the input stages have completed. I'm not sure if that totally answers your question but hopefully sheds a little more light on this design.

Reply:
I think I understand what you are saying.

For what it is worth, I think this code will intersect with the async executor (as with the async mode we'll already have a "scheduler" in something that maps works to threads, assuming we break up the work into tasks.

We may want a scheduler that prevents too much of the plan starting (aka not starting all stages at once) but I am bullish about the async approach!